core/src/main/scala/com/github/mrpowers/spark/fast/tests/ProductUtil.scala

```diff
@@ -1,35 +1,49 @@
 package com.github.mrpowers.spark.fast.tests
 
 import com.github.mrpowers.spark.fast.tests.ufansi.Color.{DarkGray, Green, Red}
+import com.github.mrpowers.spark.fast.tests.ufansi.FansiExtensions.StrOps
 import org.apache.commons.lang3.StringUtils
 import org.apache.spark.sql.Row
-import com.github.mrpowers.spark.fast.tests.ufansi.FansiExtensions.StrOps
-object DataframeUtil {
 
-  private[mrpowers] def showDataframeDiff(
+import scala.reflect.ClassTag
+
+object ProductUtil {
+  private[mrpowers] def productOrRowToSeq(product: Any): Seq[Any] = {
+    product match {
+      case r: Row     => r.toSeq
+      case p: Product => p.productIterator.toSeq
+      case null       => Seq.empty
+      case s          => Seq(s)
+    }
+  }
+  private[mrpowers] def showProductDiff[T: ClassTag](
       header: (String, String),
-      actual: Seq[Row],
-      expected: Seq[Row],
+      actual: Seq[T],
+      expected: Seq[T],
       truncate: Int = 20,
-      minColWidth: Int = 3
+      minColWidth: Int = 3,
+      defaultVal: T = null.asInstanceOf[T]
   ): String = {
 
+    val runTimeClass                     = implicitly[ClassTag[T]].runtimeClass
+    val (className, lBracket, rBracket)  = if (runTimeClass == classOf[Row]) ("", "[", "]") else (runTimeClass.getSimpleName, "(", ")")
+    val prodToString: Seq[Any] => String = s => s.mkString(s"$className$lBracket", ",", rBracket)
+    val emptyProd                        = "MISSING"
+
     val sb = new StringBuilder
 
-    val fullJoin = actual.zipAll(expected, Row(), Row())
+    val fullJoin = actual.zipAll(expected, defaultVal, defaultVal)
```

Comment:
Do we have previously a length comparator? If they are not the same length this will throw a null pointer exception, because approximateSameElements will return a false.

Reply:
No we don't since we verify schema equality beforehand so if Schema have different number of field, it should already throw the error there.

Reply:
this zip is not between elements of the same row, it's a zip between the seq of the expected and the actual dataframe/set. The rows we have from line 37 and on as `actualRow` and `expectedRow`.
If you can test this and it's ok we don't need anything.

Reply:
Ah my bad. you are right if someone use this to display DataFrame and put in a unequal Row we might have problems. I simplify the function and removed that parameter. We now display all missing row as "MISSING"